server/src/main/java/org/elasticsearch/script/VectorScoreScriptUtils.java

```diff
@@ -18,24 +19,73 @@
 public class VectorScoreScriptUtils {
 
     public static class DenseVectorFunction {
-        final ScoreScript scoreScript;
-        final float[] queryVector;
-        final DenseVectorDocValuesField field;
+        protected final ScoreScript scoreScript;
+        protected final DenseVectorDocValuesField field;
 
-        public DenseVectorFunction(ScoreScript scoreScript, List<Number> queryVector, String fieldName) {
-            this(scoreScript, queryVector, fieldName, false);
+        public DenseVectorFunction(ScoreScript scoreScript, DenseVectorDocValuesField field) {
+            this.scoreScript = scoreScript;
+            this.field = field;
+        }
+
+        void setNextVector() {
+            try {
+                field.setNextDocId(scoreScript._getDocId());
+            } catch (IOException e) {
+                throw ExceptionsHelper.convertToElastic(e);
+            }
+            if (field.isEmpty()) {
+                throw new IllegalArgumentException("A document doesn't have a value for a vector field!");
+            }
+        }
+    }
+
+    public static class ByteDenseVectorFunction extends DenseVectorFunction {
+        protected final byte[] queryVector;
+        protected final float qvMagnitude;
+
+        /**
+         * Constructs a dense vector function used for byte-sized vectors.
+         *
+         * @param scoreScript The script in which this function was referenced.
+         * @param field The vector field.
+         * @param queryVector The query vector.
+         */
+        public ByteDenseVectorFunction(ScoreScript scoreScript, DenseVectorDocValuesField field, List<Number> queryVector) {
+            super(scoreScript, field);
+            DenseVector.checkDimensions(field.get().getDims(), queryVector.size());
+
+            float[] vector = new float[queryVector.size()];
+            this.queryVector = new byte[queryVector.size()];
+            int queryMagnitude = 0;
+            for (int i = 0; i < queryVector.size(); i++) {
+                float value = queryVector.get(i).floatValue();
+                vector[i] = value;
+                this.queryVector[i] = (byte) value;
+                queryMagnitude += value * value;
+            }
+            this.qvMagnitude = (float) Math.sqrt(queryMagnitude);
+            field.getElementType().checkVectorBounds(vector);
```

Comment:
So the design here I think works pretty well because for the normal use case users already use the class bindings methods which cache the query vector as part of the script instance (which amounts to per segment). And since the query vector is a `List<Number>` we can just check it's actually a byte and then use cache a `byte[]`. This becomes a lot more complex and less performant if users want to operate on the `DenseVector` directly.